include/MeshKernel/CurvilinearGridAlgorithm.hpp

```diff
@@ -0,0 +1,71 @@
+//---- GPL ---------------------------------------------------------------------
+//
+// Copyright (C)  Stichting Deltares, 2011-2021.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+// contact: delft3d.support@deltares.nl
+// Stichting Deltares
+// P.O. Box 177
+// 2600 MH Delft, The Netherlands
+//
+// All indications and logos of, and references to, "Delft3D" and "Deltares"
+// are registered trademarks of Stichting Deltares, and remain the property of
+// Stichting Deltares. All rights reserved.
+//
+//------------------------------------------------------------------------------
+
+#pragma once
+
+#include <memory>
+
+#include <MeshKernel/CurvilinearGrid.hpp>
+#include <MeshKernel/CurvilinearGridLine.hpp>
+#include <MeshKernel/Entities.hpp>
+#include <MeshKernel/Splines.hpp>
+
+namespace meshkernel
+{
+    class CurvilinearGrid;
+
+    /// @brief A class implementing the curvilinear grid orthogonalization algorithm
+    class CurvilinearGridAlgorithm
+    {
+    public:
+        /// @brief Class constructor
+        /// @param[in] grid                        The input curvilinear grid
+        CurvilinearGridAlgorithm(std::shared_ptr<CurvilinearGrid> grid);
+
+        /// @brief Executes the algorithm, pure virtual
+        virtual std::shared_ptr<CurvilinearGrid> Compute() = 0;
```

Comment:
sometimes this function returns a copy of the input grid (like in CurvilinearGridLineShift), and sometimes it edits the input grid (lime in CurvilinearGridOrthogonalization)and returns the original  but modified input grid. That can be a bit confusing.

Reply:
Thanks. There are two approaches I think, One is conceiving the algorithms as modifiers of the grid instance (the approach i tried to follow). Another is to make a deep copy of the grid instance, where the algorithm can operate. In the end, this copy is returned from computing. Sometimes we need such a deep copy, that's why I struggled to have a uniform approach across all algorithm implementations. I now take the second approach, where a deep copy is made, assuming the cost of this deep copy is not too expensive. tests will tell